Optimize ConfigureFonts and Logging during startup

eerhardt opened the pull request:
Looking at a profile, ConfigureFonts is taking over 7ms out of 445ms of starting a `dotnet new maui` app. You can see from this speedscope trace that there are 2 things inside it taking time:

![image](https://user-images.githubusercontent.com/8291187/157109166-9e2de3dc-f8ce-4b47-b1b6-38007d49e8eb.png)

1. CreateLogger
2. GetTempPath

Optimizing the startup path so this is minimalized.
- Don't create Loggers until they are needed
- Don't use the built-in logging infrastructure unless someone explicitly enables logging
- Delay getting the TempPath in Android's EmbeddedFontLoader until it is needed, since it isn't needed by default
- Don't use ILoggerFactory to create a generic logger. Instead get the ILogger<T> service directly, so it is cached.

eerhardt committed:
Optimize Logging during startup

- Don't create Loggers until they are needed
- Don't use the built-in logging infrastructure unless someone explicitly enables logging
- Delay getting the TempPath in Android's EmbeddedFontLoader until it is needed
- Don't use ILoggerFactory to create a generic logger. Instead get the ILogger<T> service directly, so it is cached.

jonathanpeppers commented on src/Core/src/Fonts/EmbeddedFontLoader.Windows.cs:
```diff
@@ -34,7 +39,7 @@ public partial class EmbeddedFontLoader
 			catch (Exception ex)
 			{
-				_logger?.LogWarning(ex, "Unable copy font {Font} to local file system.", font.FontName);
+				_serviceProvider?.CreateLogger<FontManager>()?.LogWarning(ex, "Unable copy font {Font} to local file system.", font.FontName);
```
Can `_serviceProvider` be non-nullable and readonly in each of these classes? So we wouldn't have to do `_serviceProvider?.Something()` when using it?

eerhardt replied:
Even though these are public, I assume we don't expect "normal" callers to create them? They should get them through DI, right?

mattleibow replied:
This may no longer be true, but is ctor arg injection slower than just this func? It has to build up and search overloads and things.

Or is the difference negligible?

eerhardt replied:
There are places that explicitly pass `null` in for these parameters. For example:

https://github.com/dotnet/maui/blob/ec33802c4ab453e13a24b1fc63bb019cd28152f9/src/Core/src/Fonts/EmbeddedFontLoader.cs#L20-L21

mattleibow replied:
Issue with iserviceprovider args is that it does not describe what you are looking for. Maybe we should just use the overload that takes a func?

eerhardt replied:
> This may no longer be true, but is ctor arg injection slower than just this func? It has to build up and search overloads and things.

In the case of Logging, initializing a Logger is what is slow. I can't really make it faster without redesigning Logging all up - there is just so many layers to it, and the fact that it uses generic types, and that its constructor uses a StringBuilder to build up a category name, etc. It is just a really slow thing to create. So if we don't need it, don't create it.

If you use ctor arg injection, then DI has to create your dependencies when it creates you. So there is no way to "lazy load" them this way.

> Maybe we should just use the overload that takes a func?

Can you explain this more?

mattleibow replied:
Use Func<ILogger<T>> in the args. It can be an overload.

eerhardt replied:
I'm not sure that would be much of an improvement. It would just be another layer of indirection, wouldn't it?

Instead of:

```C#
builder.Services.TryAddSingleton<IEmbeddedFontLoader>(svc => new EmbeddedFontLoader(svc));
```

it would be

```C#
builder.Services.TryAddSingleton<IEmbeddedFontLoader>(svc => new EmbeddedFontLoader(() => svc.CreateLogger<FontManager>()));
```

And we would still need to check it for `null`. So at the log site we would have:

```C#
_getLogger?.Invoke()?.LogWarning(ex, "Unable copy font {Font} to local file system.", font.FontName);
```

mattleibow replied:
I mean we could cache it on first use. This way the type is clear and also it is cached. Maybe there is a `Logger` property? In fact, we _should_ cache the logger, right? 

And if we always have a `NullLogger`, then this will never be null. But, I can't imagine that a null check is the real perf hit here.

mattleibow replied:
Basically, my main concern is that it takes `IServiceProvider` and that is a bit of a no-no in the DI world, but also it doesn't let the callers and mocking system know what to provide. Maybe we should have an overload - so both the type and the ISP

Right now, I don't think our tests actually check to see if there was an internal exception, and one way is the logger.

I am just scared of the opaque ISP... Maybe I am just pessimistic. If nobody else shares this concern, then maybe it is not a real issue and I am just a scared guy XD

PureWeen replied:
> is a bit of a no-no in the DI world, but also it doesn't let the callers and mocking system know what to provide. Maybe we should have an overload - so both the type and the ISP

I'd say it's a no-no with exceptions and that lot of that advice originated from scenarios that weren't trying to make an Android app start up quickly. 

As long as we don't start resolving 10s of dependencies through service locator then it's fine. 

If we want to have clear dependencies on the `FontManager` maybe we could register a singleton `IMauiLoggerFactory` and the concrete implementation of that takes an `IServiceProvider` on the `ctor`. Then the fuzziness of ISP is isolated to the factory implementation? Though maybe that much indirection will just start to ruin the benefits here.

eerhardt replied:
Just so I fully understand the concerns here, are the concerns with taking IServiceProvider here basically the same that are listed in:

https://blog.ploeh.dk/2015/10/26/service-locator-violates-encapsulation/
https://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/

jonathanpeppers commented on src/Core/src/Fonts/EmbeddedFontLoader.Android.cs:
```diff
@@ -1,8 +1,22 @@
 	public partial class EmbeddedFontLoader : FileSystemEmbeddedFontLoader
 	{
+		public EmbeddedFontLoader(IServiceProvider? serviceProvider = null)
+			: base(GetTempPath, serviceProvider)
```
The difference between the Android API and `System.IO.Path.GetTempPath()` appears to be a trailing slash:
```
03-07 14:18:36.776 19754 19754 I DOTNET  : Path: /data/user/0/com.companyname.foo/cache/
03-07 14:18:36.787 19754 19754 I DOTNET  : Android: /data/user/0/com.companyname.foo/cache
```
Maybe let's just change this to this?
```suggestion
			: base(Path.GetTempPath, serviceProvider)
```
And delete `GetTempPath()`?

eerhardt replied:
I'm not sure I feel comfortable changing the behavior here. Are we certain that your suggestion is safe?

jonathanpeppers replied:
If they are looking for a temp folder to extract fonts to, it seems like `Path.GetTempPath()` is what they would want? I don't think you need to call into Java to get this path.

I think they would only use this folder if a font is an `EmbeddedResource`, because there isn't an Android API for loading a font from a stream.

eerhardt replied:
Looks like this code was added in https://github.com/dotnet/maui/pull/1352, which says:

> Avoid the use of Path.GetTempPath() in the font loader because the first access to IO is slow. move it on a bit and see what changes

eerhardt replied:
cc @mattleibow

mattleibow replied:
I got this info from @grendel and we changed it because it was expensive.

eerhardt replied:
So we no longer use the filesystem in the font startup path anymore. So if this was just added for perf reasons, can I change it back to just `Path.GetTempPath`?

mattleibow replied:
Yeah, that makes sense. Keep it nice and clean.

eerhardt committed:
Fix unit tests to always create a new UnitTestLoggerProvider, even if the test doesn't initialize an ILogger. This way logs from previous tests are cleared.

mattleibow commented on src/Core/src/Hosting/ServiceProviderExtensions.cs:
```diff
@@ -11,14 +11,12 @@ internal static class ServiceProviderExtensions
 			context.Services.CreateLogger<T>();
 
 		internal static ILogger<T>? CreateLogger<T>(this IServiceProvider services) =>
-			services.GetService<ILoggerFactory>()?.CreateLogger<T>();
```
This extension method does not look very helpful  anymore, maybe we should just remove it.

eerhardt replied:
I think I'm going to keep it for now to make it symmetrical with the other methods here - ex. extension on MauiContext.

We can maybe clean this up later, if needed. It's all internal anyway.

eerhardt committed:
PR feedback

- Remove the call to the Android API to get the temp path and just call Path.GetTempPath
